File: packages/vscode/src/node-locator.ts
import { NodeNotFound } from "@groks-beard/core"
import { Effect } from "effect"
import { resolveSpawnTarget } from "./cli-locator.js"

export type LocateNodeInput = {
  readonly nodePath?: string
  readonly env: Record<string, string | undefined>
  readonly exists: (path: string) => boolean
  readonly win?: boolean
}

const join = (dir: string, name: string): string => `${dir.replace(/[\\/]+$/, "")}/${name}`

export const nodeBinaryName = (win: boolean): string => (win ? "node.exe" : "node")

export const isAbsoluteNodePath = (path: string, win: boolean): boolean => {
  if (path === "") return false
  if (win) return /^[a-zA-Z]:[\\/]/.test(path) || path.startsWith("\\\\")
  return path.startsWith("/")
}

export const candidateNodePaths = (input: LocateNodeInput): ReadonlyArray<string> => {
  const win = input.win ?? false
  const bin = nodeBinaryName(win)
  const out: Array<string> = []
  if (input.nodePath !== undefined && isAbsoluteNodePath(input.nodePath, win)) {
    out.push(input.nodePath)
  }
  const pathEnv = input.env.PATH ?? input.env.Path ?? ""
  const sep = win ? ";" : ":"
  for (const dir of pathEnv.split(sep)) {
    if (dir.trim() === "") continue
    out.push(join(dir, bin))
    if (win) out.push(join(dir, "node.cmd"))
  }
  return out
}

export const locateNode = (input: LocateNodeInput): Effect.Effect<string, NodeNotFound> => {
  const searched = candidateNodePaths(input)
  const hit = searched.find((path) => input.exists(path))
  if (hit === undefined) return Effect.fail(new NodeNotFound({ searched: [...searched] }))
  return Effect.succeed(resolveSpawnTarget(hit, input.exists))
}

File: beard/core/src/main/scala/groksbeard/core/NodeLocator.scala
package groksbeard.core

final case class LocateNode(
    nodePath: Option[String],
    pathEnv: String,
    win: Boolean,
    exists: String => Boolean,
)

object NodeLocator:
  def binaryName(win: Boolean): String = if win then "node.exe" else "node"

  def isAbsolute(path: String, win: Boolean): Boolean =
    if path.isEmpty then false
    else if win then path.matches("^[a-zA-Z]:[\\\\/].*") || path.startsWith("\\\\")
    else path.startsWith("/")

  def join(dir: String, name: String): String =
    s"${dir.replaceAll("[\\\\/]+$", "")}/$name"

  def candidates(input: LocateNode): List[String] =
    val bin = binaryName(input.win)
    val out = List.newBuilder[String]
    input.nodePath.filter(p => isAbsolute(p, input.win)).foreach(out += _)
    val sep = if input.win then ";" else ":"
    input.pathEnv.split(sep).foreach { dir =>
      if dir.trim.nonEmpty then
        out += join(dir, bin)
        if input.win then out += join(dir, "node.cmd")
    }
    out.result()
  end candidates

  def resolveSpawnTarget(candidate: String, exists: String => Boolean): String =
    if candidate.matches("(?i).*\\.(cmd|bat)$") then
      val exe = candidate.replaceAll("(?i)\\.(cmd|bat)$", ".exe")
      if exists(exe) then exe else candidate
    else candidate

  def locate(input: LocateNode): Either[List[String], String] =
    val searched = candidates(input)
    searched.find(input.exists) match
      case None      => Left(searched)
      case Some(hit) => Right(resolveSpawnTarget(hit, input.exists))
end NodeLocator
